### PR TITLE
Add api white list and rate limiter for gc

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: Ia7ac5dc5d7d91676701738f2a0ae5073f9ceb01e

### CODE REVIEW

File: hugegraph-example/src/main/resources/hugegraph.properties [hunk 1]
```diff
@@ -1,10 +1,10 @@
 gremlin.graph=com.baidu.hugegraph.HugeFactory
 
-backend=cassandra
-serializer=cassandra
+#backend=cassandra
```
Comment (zhoney) [inline hugegraph-example/src/main/resources/hugegraph.properties]: is expected？

Comment (Linary) [reply hugegraph-example/src/main/resources/hugegraph.properties]: Yeah, it's convenient for local test

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java [hunk 2]
```diff
@@ -69,4 +87,15 @@ public void filter(ContainerRequestContext context) {
                       ServerOptions.MAX_WORKER_THREADS.name()));
         }
     }
+
+    private static boolean isWhiteAPI(ContainerRequestContext context) {
+        String rootPath = context.getUriInfo().getPathSegments().get(0).getPath();
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java]: check size of context.getUriInfo().getPathSegments() > 0

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java [hunk 3]
```diff
@@ -19,38 +19,59 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.RateLimiter;
 
 @Provider
 @Singleton
 @PreMatching
 public class LoadDetectFilter implements ContainerRequestFilter {
 
+    private static final Set<String> WHITE_API_LIST = ImmutableSet.of(
+            "metrics",
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/filter/LoadDetectFilter.java]: add root path ""

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
+/*
+ *
+ *  * Copyright 2017 HugeGraph Authors
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java]: there are tow "*"

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        profiles.put("apis", apis);
+        SERVER_PROFILES = JsonUtil.toJson(profiles);
+        return SERVER_PROFILES;
+    }
+
+    @GET
+    @Path("apis")
+    @Timed
+    @Produces(MediaType.APPLICATION_JSON)
+    public String showAll(@Context Application application) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java]: showAllAPIs

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        if (SERVER_PROFILES != null) {
+            return SERVER_PROFILES;
+        }
+
+        Map<String, Object> profiles = InsertionOrderUtil.newMap();
+        profiles.put("service", SERVICE);
+        profiles.put("doc", DOC);
+        profiles.put("api-doc", API_DOC);
+        Set<String> apis = new HashSet<>();
+        for (Class<?> aClass : application.getClasses()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java]: Class<?> clazz

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        API_PROFILES = JsonUtil.toJson(apiProfiles);
+        return API_PROFILES;
+    }
+
+    private static boolean isAnnotatedPathClass(Class rc) {
+        if (rc.isAnnotationPresent(Path.class)) {
+            return true;
+        }
+        for (Class i : rc.getInterfaces()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    @Path("apis")
+    @Timed
+    @Produces(MediaType.APPLICATION_JSON)
+    public String showAll(@Context Application application) {
+        if (API_PROFILES != null) {
+            return API_PROFILES;
+        }
+
+        Map<String, Map<String, List<APIProfile>>> apiProfiles = new HashMap<>();
+        for (Class<?> aClass : application.getClasses()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/ProfileAPI.java]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/DataTypeTest.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Test
+    public void testValueToNumber() {
+        Assert.assertNull(DataType.BOOLEAN.valueToNumber(1));
+        Assert.assertNull(DataType.INT.valueToNumber("not number"));
+
+        Assert.assertEquals((byte) 1, DataType.BYTE.valueToNumber(1));
+        Assert.assertEquals(1, DataType.INT.valueToNumber(1));
+        Assert.assertEquals(1, DataType.INT.valueToNumber(1.0F));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/DataTypeTest.java]: Assert.assertEquals(1, DataType.INT.valueToNumber((byte) 1));

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/DataTypeTest.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        Assert.assertEquals("text", DataType.TEXT.string());
+        Assert.assertEquals("blob", DataType.BLOB.string());
+        Assert.assertEquals("date", DataType.DATE.string());
+        Assert.assertEquals("uuid", DataType.UUID.string());
+    }
+
+    @Test
+    public void testValueToNumber() {
+        Assert.assertNull(DataType.BOOLEAN.valueToNumber(1));
+        Assert.assertNull(DataType.INT.valueToNumber("not number"));
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/DataTypeTest.java]: Assert.assertNull(DataType.INT.valueToNumber(1.0f));

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/ProfileAPI.java [hunk 11]
```diff
@@ -43,6 +43,8 @@
 import com.baidu.hugegraph.util.E;
 import com.baidu.hugegraph.util.InsertionOrderUtil;
 import com.baidu.hugegraph.util.JsonUtil;
+import com.baidu.hugegraph.version.ApiVersion;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/ProfileAPI.java]: unused import

**Pull request merged**